aufs optimizations

kolyshkin opened the pull request:
1. Remove global Get()/Put() locking, add lock around mount().
2. Streamline Put() (unmount).
3. Optimize lots-of-layers case.
3. Simplify Cleanup().

For detailed description, see commit messages.

### Performance and stress testing

This was tested by the following script on a machine with 8GB RAM (with more RAM you can increase the value of `PARALLEL`):

```bash
#!/bin/bash

NUMCTS=1000
PARALLEL=100
IMAGE=golang

create() {
	# -v 'vol_{}:/vol'
	time ( seq $NUMCTS | parallel -j$PARALLEL docker create $IMAGE top > /dev/null )
	echo "^^^ CREATE TIME ^^^"
}

remove() {
	time ( docker ps -qa | shuf | tail -n $NUMCTS | parallel -j$PARALLEL docker rm -f '{}' > /dev/null )
	echo "^^^ REMOVE TIME ^^^"
}

## MAIN

docker pull $IMAGE

# stage 1: precreate $NUMCTS containers
create
echo

# stage2: create another $NUMCTS while removing $NUMCTS containers
create &
remove &
wait

# stage 3: remove the rest of it
remove
```

Results for the above tests, measured at least 3 times (with https://github.com/moby/moby/pull/39135 applied, with and without this PR):
 - container creation on stage 1 is 3x faster (from 60s to 20s);
 - container creation while removal (stage 2) is almost 2x faster (from 85s to 45s);
 - container removal (stages 2 and 3) times are about the same;

No errors were detected.

kolyshkin committed:
aufs: remove extra locking

Both mount and unmount calls are already protected by fine-grained
(per id) locks in Get()/Put() introduced in commit fc1cf1911bb
("Add more locking to storage drivers"), so there's no point in
having a global lock in mount/unmount.

The only place from which unmount is called without any locking
is Cleanup() -- this is to be addressed in the next patch.

This reverts commit 824c24e6802ad3ed7e26b4f16e5ae81869b98185.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
aufs: use mount.Unmount

1. Use mount.Unmount() which ignores EINVAL ("not mounted") error,
and provides better error diagnostics (so we don't have to explicitly
add target to error messages).

2. Since we're ignoring "not mounted" error, we can call
multiple unmounts without any locking -- but since "auplink flush"
is still involved and can produce an error in logs, let's keep
the check for fs being mounted (it's just a statfs so should be fast).

2. While at it, improve the "can't unmount" error message in Put().

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
aufs.Cleanup: optimize

Do not use filepath.Walk() as there's no requirement to recursively
go into every directory under mnt -- a (non-recursive) list of
directories in mnt is sufficient.

With filepath.Walk(), in case some container will fail to unmount,
it'll go through the whole container filesystem which is both
excessive and useless.

This is similar to commit f1a459229724f5e8e440b49f ("devmapper.shutdown:
optimize")

While at it, raise the priority of "unmount error" message from debug
to a warning. Note we don't have to explicitly add `m` as unmount error (from
pkg/mount) will have it.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented:
```
03:02:27 daemon/graphdriver/aufs/aufs.go:604:11:warning: undeclared name: x (gosimple)
03:02:27 daemon/graphdriver/aufs/aufs.go:426:11:warning: a.unmount undefined (type *Driver has no field or method unmount) (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:566:21:warning: cannot use m (variable of type os.FileInfo) as string value in argument to Unmount (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:604:11:warning: undeclared name: x (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:314:6:warning: undeclared name: err (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:313:3:warning: undeclared name: err (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:313:11:warning: a.unmount undefined (type *Driver has no field or method unmount) (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:322:24:warning: undeclared name: err (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:319:24:warning: undeclared name: err (unconvert)
03:02:27 daemon/graphdriver/aufs/aufs.go:318:19:warning: undeclared name: err (unconvert)
```

thaJeztah commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -311,27 +310,16 @@ func (a *Driver) Remove(id string) error {
-			break
-		}
-
 		err = a.unmount(mountpoint)
```
```suggestion
		err := a.unmount(mountpoint)
```

kolyshkin replied:
Sorry for the noise, have pushed some intermediate (unfinished) version late last night... should be fixed now.

thaJeztah commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -632,7 +601,7 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
 	}
 	data := label.FormatMountLabel(fmt.Sprintf("%s,%s", string(b[:bp]), opts), mountLabel)
-	if err = unix.Mount("none", target, "aufs", 0, data); err != nil {
+	if err = x.Mount("none", target, "aufs", 0, data); err != nil {
```
```suggestion
	if err = unix.Mount("none", target, "aufs", 0, data); err != nil {
```

kolyshkin committed:
aufs: aufsMount: better errors for unix.Mount()

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

cpuguy83 commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -579,23 +561,20 @@ func (a *Driver) mounted(mountpoint string) (bool, error) {
 // Cleanup aufs and unmount all mountpoints
 func (a *Driver) Cleanup() error {
-	var dirs []string
-	if err := filepath.Walk(a.mntPath(), func(path string, info os.FileInfo, err error) error {
```
😱

kolyshkin replied:
We have something similar in devmapper, too...

cpuguy83 commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -579,23 +561,20 @@ func (a *Driver) mounted(mountpoint string) (bool, error) {
+	dir := a.mntPath()
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return errors.Wrap(err, "aufs.Cleanup() failed")
```
How about `error reading aufs mount dir`

kolyshkin replied:
fixed

cpuguy83 reviewed (commented):
I'm pretty sure I tried to remove this lock before and ran into trouble.

kolyshkin committed:
aufs: add lock around mount

Apparently there is some kind of race in aufs kernel module code,
which leads to the errors like:

[98221.158606] aufs au_xino_create2:186:dockerd[25801]: aufs.xino create err -17
[98221.162128] aufs au_xino_set:1229:dockerd[25801]: I/O Error, failed creating xino(-17).
[98362.239085] aufs au_xino_create2:186:dockerd[6348]: aufs.xino create err -17
[98362.243860] aufs au_xino_set:1229:dockerd[6348]: I/O Error, failed creating xino(-17).
[98373.775380] aufs au_xino_create:767:dockerd[27435]: open /dev/shm/aufs.xino(-17)
[98389.015640] aufs au_xino_create2:186:dockerd[26753]: aufs.xino create err -17
[98389.018776] aufs au_xino_set:1229:dockerd[26753]: I/O Error, failed creating xino(-17).
[98424.117584] aufs au_xino_create:767:dockerd[27105]: open /dev/shm/aufs.xino(-17)

So, we have to have a lock around mount syscall.

While at it, don't call the whole Unmount() on an error path, as
it leads to bogus error from auplink flush.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
aufs: optimize lots of layers case

In case there are a big number of layers, so that mount data won't fit
into a single memory page (4096 bytes on most platforms, which is good
enough for about 40 layers, depending on how long graphdriver root path
is), we supply additional layers with O_REMOUNT, as described in aufs
documentation.

Problem is, the current implementation does that one layer at a time
(i.e. there is one mount syscall per each additional layer).

Optimize the code to supply as many layers as we can fit in one page
(basically reusing the same code as for the original mount).

Note, per aufs docs, "[a]t remount-time, the options are interpreted
in the given order, e.g. left to right" so we should be good.

Tested on an image with ~100 layers.

Before (35 syscalls):
> [pid 22756] 1556919088.686955 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/a86f8c9dd0ec2486293119c20b0ec026e19bbc4d51332c554f7cf05d777c9866", "aufs", 0, "br:/mnt/volume_sfo2_09/docker-au"...) = 0 <0.000504>
> [pid 22756] 1556919088.687643 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/a86f8c9dd0ec2486293119c20b0ec026e19bbc4d51332c554f7cf05d777c9866", 0xc000c451b0, MS_REMOUNT, "append:/mnt/volume_sfo2_09/docke"...) = 0 <0.000105>
> [pid 22756] 1556919088.687851 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/a86f8c9dd0ec2486293119c20b0ec026e19bbc4d51332c554f7cf05d777c9866", 0xc000c451ba, MS_REMOUNT, "append:/mnt/volume_sfo2_09/docke"...) = 0 <0.000098>
> ..... (~30 lines skipped for clarity)
> [pid 22756] 1556919088.696182 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/a86f8c9dd0ec2486293119c20b0ec026e19bbc4d51332c554f7cf05d777c9866", 0xc000c45310, MS_REMOUNT, "append:/mnt/volume_sfo2_09/docke"...) = 0 <0.000266>

After (2 syscalls):
> [pid 24352] 1556919361.799889 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/8e7ba189e347a834e99eea4ed568f95b86cec809c227516afdc7c70286ff9a20", "aufs", 0, "br:/mnt/volume_sfo2_09/docker-au"...) = 0 <0.001717>
> [pid 24352] 1556919361.801761 mount("none", "/mnt/volume_sfo2_09/docker-aufs/aufs/mnt/8e7ba189e347a834e99eea4ed568f95b86cec809c227516afdc7c70286ff9a20", 0xc000dbecb0, MS_REMOUNT, "append:/mnt/volume_sfo2_09/docke"...) = 0 <0.001358>

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
> I'm pretty sure I tried to remove this lock before and ran into trouble.

@cpuguy83 yeah, it took me quite a while (days) to figure out what the proper locking should be (and I have yet to update this PR with my new patchset). Apparently, parallel unmounts is fine, but there's a race in aufs kernel module that prevents parallel mounts, so this has to be done under a lock.

Will try to bring this back to life today.

kolyshkin commented:
OK, I have updated the patch set and did some benchmarks (on top of #39135, using the script from there). Parallel container creation is now twice faster, other timings are similar.

thaJeztah commented:
@kolyshkin build is failing;

```
18:54:56 # github.com/docker/docker/daemon/graphdriver/aufs
18:54:56 daemon/graphdriver/aufs/aufs.go:626: Sprintf format %s reads arg #1, but call has 0 args
18:54:57 FAIL	github.com/docker/docker/daemon/graphdriver/aufs [build failed]
```

thaJeztah commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -632,14 +612,29 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
+	for index < len(ro) {
+		bp = 0
+		for ; index < len(ro); index++ {
+			layer := fmt.Sprintf("append:%s=ro+wh,")
```
Looks like it's failing here

kolyshkin replied:
as always, I pushed some older version. fixed now.

kolyshkin commented:
Not sure if this is relevant (error on exp CI):

> 00:14:27.831 --- FAIL: TestServiceCreateInit (5.81s)
> 00:14:27.831     --- FAIL: TestServiceCreateInit/daemonInitDisabled (2.45s)
> 00:14:27.831         create_test.go:38: Creating a new daemon
> 00:14:27.831         daemon.go:288: [de58392204bd7] waiting for daemon to start
> 00:14:27.831         daemon.go:320: [de58392204bd7] daemon started
> 00:14:27.831         create_test.go:52: assertion failed: error is not nil: Error response from daemon: rpc error: code = AlreadyExists desc = name conflicts with an existing object: service gracious_ramanujan already exists: error creating service
> 00:14:27.831         daemon.go:275: [de58392204bd7] exiting daemon

thaJeztah commented:
perhaps the name generator wasn't random enough? Interesting; never encountered that

kolyshkin commented:
janky CI is flaky test (https://github.com/moby/moby/issues/32673)
> 01:04:57.140 FAIL: docker_api_swarm_test.go:292: DockerSwarmSuite.TestAPISwarmLeaderElection

kolyshkin commented:
adding a temp commit to stress-test TestService (run them many times)

kolyshkin commented:
Also checking the same w/o my patches at https://github.com/moby/moby/pull/39171

kolyshkin commented:
Windows CI failure, def unrelated (as this PR changes aufs graph driver used by Linux only)
```
00:39:15.779 
00:39:15.779 ----------------------------------------------------------------------
00:39:15.779 FAIL: check_test.go:107: DockerSuite.TearDownTest
00:39:15.779 
00:39:15.779 assertion failed: error is not nil: Error response from daemon: container ddb0019e76d456354c4761d6426da149fe516d1eb437737e4a1fc26ef03f4b67: driver "windowsfilter" failed to remove root filesystem: hcsshim::GetComputeSystems: The requested compute system operation is not valid in the current state.: failed to remove ddb0019e76d456354c4761d6426da149fe516d1eb437737e4a1fc26ef03f4b67
00:39:15.779 
```

kolyshkin commented:
I am seeing TestService* test timeouts after 5m when run as
```
export TEST_REPEAT=10
export TESTFLAGS="-test.count 10 -test.run TestService"
./hack/make.sh test-integration
```
even without the patches from this PR (see https://github.com/moby/moby/pull/39171), so this PR is not the one to blame.

kolyshkin commented:
OK, I am really doing something stupid here. CI does not test aufs driver, so its results are irrelevant.

kolyshkin commented:
> 01:28:27.996 FAIL: docker_cli_swarm_test.go:1317: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

is a known flaky test; this is the only CI failure

thaJeztah commented:
@cpuguy83 @dmcgowan ptal

cpuguy83 reviewed (commented):
I would like to get a chance to give this a thorough manual test.
I recall before seeing issues with the auplink call.

kolyshkin commented:
> I would like to get a chance to give this a thorough manual test.

Yes please; I hope you can think of something else than the test I come up with (it's in the description of this thread). If you'll run mine, make sure to change PARALLEL value to suit your RAM to avoid ENOMEM -- in my case 100 is good for 8GB, 150 occasionally gives ENOMEM.

> I recall before seeing issues with the auplink call.

The issue I was seeing was a race in the aufs kernel module, something about xino -- so I ended up not allowing more than a single `mount -t aufs` at a time.

The only issue I saw with `auplink flush` is when it's called on an error path where the aufs mount failed, and it fails because there's no mount. But maybe my testing is just not good enough.

kolyshkin commented:
NOTE: I just modified one of the commits (a0d0b88b5cc1729e "aufs: use mount.Unmount") so it no longer removes a `mounted` check from Remove() method.

cpuguy83 commented:
So there is a benchmark in aufs test dir for concurrent access.
WIth this change I see a bunch of:

```
WARN[0001] Couldn't run auplink before unmount /tmp/aufs-tests/aufs/mnt/3ed1ff0d4927f4741265c6eefb3859d9bccd47c919ec5b053a3c380fcd4da6ce  error="exit status 22" storage-driver=aufs
```

dmcgowan commented:
I can see how the global lock is unnecessary, but the lock should at least be on the target, which this PR is not doing. I don't understand what putting a lock around `unix.Mount` is accomplishing since there is no additional check after the lock is acquired.

kolyshkin commented:
@cpuguy83 thanks, I'll take a look (I also confirm that there are no such warnings before this patch).

> but the lock should at least be on the target, which this PR is not doing

@dmcgowan quoting my commit message
>    Both mount and unmount calls are already protected by fine-grained
>    (per id) locks in Get()/Put() introduced in commit fc1cf1911bb
>    ("Add more locking to storage drivers"), so there's no point in
>    having a global lock in mount/unmount.

So there is per-id (which can be though of a short for mount point) locking in place already. Let me know if you think it's not adequate, or say the lock should really be on a per mount-point rather than just id.

> I don't understand what putting a lock around unix.Mount is accomplishing
@dmcgowan as described in patch commit message, I have discovered that doing two mounts in parallel (with different mount points and arguments, i.e. unrelated ones) sometimes hit a bug in the aufs kernel module (something concerning xino handling, see the commit for more details), and so I had to introduce a lock around doing a mount() syscall so no two mounts happen concurrently.

kolyshkin commented:
The error from `auplink flush` during benchmark is
> auplink:proc_mnt.c:96: /tmp/aufs-tests/aufs/mnt/36f594c070a6576ebc1a952da4c1d493067c7b134a979d636dfb6fe34e65993c: Invalid argument

comes from https://sourceforge.net/p/aufs/aufs-util/ci/aufs3.2/tree/proc_mnt.c#l96

and means that auplink can't find this mount in /proc/self/mounts.

kolyshkin commented:
> and means that auplink can't find this mount in /proc/self/mounts.

My theory is this happens because the kernel can not guarantee the continuity of /proc/self/mounts contents, i.e. when many mounts are added and removed, while someone is reading /proc/self/mounts, its contents might be broken. Let me try to prove it...

kolyshkin commented:
> My theory is this happens because the kernel can not guarantee the continuity of /proc/self/mounts contents

Confirmed by some tests; in particular, grep before and after auplink flush shows that the mount entry is there, yet auplink sometimes fails to find it.

Talked to @avagin and he confirmed it might be the case with the current kernel.

Noticed that `getmntent_r` reads the file by 1K blocks, while entries are close to 4K in size, so each line requires 4 reads. Making the buffer 4K in auplink doesn't help though.

What helps is to retry auplink flush :-\ and while I hate such solutions, I don't know what else to do (except to rewrite auplink or fix the kernel).

kolyshkin committed:
aufs: retry auplink flush

Running a bundled aufs benchmark sometimes results in this warning:

> WARN[0001] Couldn't run auplink before unmount /tmp/aufs-tests/aufs/mnt/XXXXX  error="exit status 22" storage-driver=aufs

If we take a look at what aulink utility produces on stderr, we'll see:

> auplink:proc_mnt.c:96: /tmp/aufs-tests/aufs/mnt/XXXXX: Invalid argument

and auplink exits with exit code of 22 (EINVAL).

Looking into auplink source code, what happens is it tries to find a
record in /proc/self/mounts corresponding to the mount point (by using
setmntent()/getmntent_r() glibc functions), and it fails.

Some manual testing, as well as runtime testing with lots of printf
added on mount/unmount, as well as calls to check the superblock fs
magic on mount point (as in graphdriver.Mounted(graphdriver.FsMagicAufs, target)
confirmed that this record is in fact there, but sometimes auplink
can't find it. I was also able to reproduce the same error (inability
to find a mount in /proc/self/mounts that should definitely be there)
using a small C program, mocking what `auplink` does:

```c
 #include <stdio.h>
 #include <err.h>
 #include <mntent.h>
 #include <string.h>
 #include <stdlib.h>

int main(int argc, char **argv)
{
	FILE *fp;
	struct mntent m, *p;
	char a[4096];
	char buf[4096 + 1024];
	int found =0, lines = 0;

	if (argc != 2) {
		fprintf(stderr, "Usage: %s <mountpoint>\n", argv[0]);
		exit(1);
	}

	fp = setmntent("/proc/self/mounts", "r");
	if (!fp) {
		err(1, "setmntent");
	}
	setvbuf(fp, a, _IOLBF, sizeof(a));
	while ((p = getmntent_r(fp, &m, buf, sizeof(buf)))) {
		lines++;
		if (!strcmp(p->mnt_dir, argv[1])) {
			found++;
		}
	}
	printf("found %d entries for %s (%d lines seen)\n", found, argv[1], lines);
	return !found;
}
```

I have also wrote a few other C proggies -- one that reads
/proc/self/mounts directly, one that reads /proc/self/mountinfo instead.
They are also prone to the same occasional error.

It is not perfectly clear why this happens, but so far my best theory
is when a lot of mounts/unmounts happen in parallel with reading
contents of /proc/self/mounts, sometimes the kernel fails to provide
continuity (i.e. it skips some part of file or mixes it up in some
other way). In other words, this is a kernel bug (which is probably
hard to fix unless some other interface to get a mount entry is added).

Now, there is no real fix, and a workaround I was able to come up
with is to retry when we got EINVAL. It usually works on the second
attempt, although I've once seen it took two attempts to go through.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@cpuguy83 please see the last commit for my findings and the proposed fix. Ugly, I know :-\

kolyshkin commented:
In a futile attempt to fix it in some other way I wrote 3 or 4 implementations of /proc/self/mounts reader, as well as /proc/self/mountinfo (in a hope that it does not have the same continuity issue). They all fail at times, so the problem is in the kernel.

I'll see if I can patch auplink to re-read mount table at least once before failing.

kolyshkin commented:
I also sent a patch to aufs mailing list and to its author as this can be fixed in auplink as well (so we won't have to call it twice). I don't see it in the mailing list yet (must be premoderated), so here's the description 
```
Date: Mon, 13 May 2019 17:04:31 -0700
Subject: [PATCH] au_proc_getmntent: optimize reading, retry

1. By looking into strace output, it looks like default buffer
for fread() (and thus getmntent()) is only 1K, while a line in
/proc/self/mounts can easily be more than 4K.
Set the libc reading buffer to 4K, as this is a page size for'
most architectures and thus reading is more efficient.

2. While doing massively parallel aufs mounts/umounts (256 or more
in parallel) I observed that sometimes auplink fails to find
a specified mountpoint in /proc/self/mounts. Presumably this happens
because the Linux kernel can't provide continuity of the file while
the mount entries are being added and removed. So far the best
solution to the problem I was able to find is to retry reading
the file -- in 100% of cases it finds the mount from the second
attempt. To be on the safe size, retry reading twice.

3. Since reading /proc/self/mount is a relatively costly operation,
try avoiding it by first checking that the mount point provided is
indeed aufs mount point (check for aufs magic). This way, we will
re-read /proc/self/mounts only if we're dealing with aufs mount.
```

cpuguy83 reviewed (approved):
LGTM

Thanks! Retry actually doesn't look bad. Definitely nice to get it into auplink but this is fine.
Excellent work!

kolyshkin commented:
Not sure whom else to ask to PTAL... @thaJeztah ?

thaJeztah commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -326,11 +326,11 @@ func (a *Driver) Remove(id string) error {
 		}
 		if retries >= 5 {
-			return errors.Wrapf(err, "aufs: unmount error after retries: %s", mountpoint)
+			return errors.Wrapf(err, "aufs: unmount error after retries")
```
```suggestion
			return errors.Wrap(err, "aufs: unmount error after retries")
```

kolyshkin replied:
fixed, thanks!

thaJeztah reviewed (approved):
SGTM

found one nit

also would like to have another maintainer give this a review (perhaps @dmcgowan @tonistiigi) ?

andrewhsu commented:
Hmm..seems to be a [consistent](https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/14427/console) [failure](https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/14422/console) on powerpc PR job only:
```
FAIL: docker_cli_search_test.go:47: DockerSuite.TestSearchCmdOptions
```

kolyshkin commented:
> Hmm..seems to be a consistent failure on powerpc PR job only:

@andrewhsu def unrelated; this PR only modifies aufs driver which is not used in these CI jobs.

This test looks poorly written :( anyway, testing it out in https://github.com/moby/moby/pull/39243

kolyshkin commented:
Rebased to include #39243

andrewhsu reviewed (approved):
SGTM w.r.t. getting perf optimizations for aufs; much desirable

kolyshkin commented:
OK, the test I wrote (https://github.com/kolyshkin/procfs-test) convinced aufs author to include my patch (that retries reading of /proc/self/mounts in case the mount can't be found). Nevertheless it makes sense to have a retry here as well, since we don't ship `auplink` utility.

kolyshkin commented:
OK, I have finally figured out why two aufs mounts might race (and so I had to add a lock around `unix.Mount()` call. The reason is xino mount option. Docker aufs graph driver have it as `xino=/dev/shm/.aufs.xino`, and the kernel (aufs module) creates the file and removes it (while keeping it open). Over the course of mount, there is a certain time window between creation and removal of this file, and if during that time the second mount will kick in, it will fail.

Aside from using lock around mount, which is what this PR does, there are other ways to work around this.

1. Use a more unique file name (such as add a suffix).
2. Do not use `xino=` attribute (in this case aufs will create the xino file on the writable branch).]
3. Fix this in aufs kernel module.

I'm going to go with approach 1.

kolyshkin commented:
I did something like this:

```diff
diff --git a/daemon/graphdriver/aufs/aufs.go b/daemon/graphdriver/aufs/aufs.go
index 9c4e28e..2fd8c8a 100644
--- a/daemon/graphdriver/aufs/aufs.go
+++ b/daemon/graphdriver/aufs/aufs.go
@@ -28,10 +28,12 @@ import (
        "fmt"
        "io"
        "io/ioutil"
+       "math/rand"
        "os"
        "os/exec"
        "path"
        "path/filepath"
+       "strconv"
        "strings"
        "sync"
        "time"
@@ -80,7 +82,6 @@ type Driver struct {
        pathCache     map[string]string
        naiveDiff     graphdriver.DiffDriver
        locker        *locker.Locker
-       mntL          sync.Mutex
 }
 
 // Init returns a new AUFS driver.
@@ -618,14 +619,13 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
                bp += copy(b[bp:], layer)
        }
 
-       opts := "dio,xino=/dev/shm/aufs.xino"
+       rnd := strconv.FormatInt(int64(1e9+rand.Uint32()%1e9), 36)[1:5]
+       opts := "dio,xino=/dev/shm/aufs."+rnd
        if useDirperm() {
                opts += ",dirperm1"
        }
        data := label.FormatMountLabel(fmt.Sprintf("%s,%s", string(b[:bp]), opts), mountLabel)
-       a.mntL.Lock()
        err = unix.Mount("none", target, "aufs", 0, data)
-       a.mntL.Unlock()
        if err != nil {
                err = errors.Wrap(err, "mount target="+target+" data="+data)
                return
@@ -641,9 +641,7 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
                        bp += copy(b[bp:], layer)
                }
                data := label.FormatMountLabel(string(b[:bp]), mountLabel)
-               a.mntL.Lock()
                err = unix.Mount("none", target, "aufs", unix.MS_REMOUNT, data)
-               a.mntL.Unlock()
                if err != nil {
                        err = errors.Wrap(err, "mount target="+target+" flags=MS_REMOUNT data="+data)
                        return
```

but it didn't change anything from the performance point of view. Not sure which way is better.
